### PR TITLE
Add hapi v17 support

### LINKED ISSUE #62: hapi v17 support

### PR DESCRIPTION
**Pull request by Samueljoli**
This resolves #62 . 

#### Breaking changes:
- `mapUri` - No longer accepts a callback and instead returns an object with the properties of `uri` and `headers` ( @hueniverse, @osukaa thoughts?)

#### Updated deps:
- `Wreck` to **v14**
- `Lab` to **v15**
- `Inert` to **v5**

### CODE REVIEW

**Comment by hueniverse** [comment]
Need to update .travis.yml file as well as figure out why we are getting these lint errors.

**Comment by Samueljoli** [comment]
@hueniverse Should we still support `argon` and `boron`? Or just `8`

**Comment by Samueljoli** [comment]
I also can't see the job logs in travis and I'm not having any linting issues locally. I'd imagine it'll resolve itself after updating `travis.yml`

**Comment by Samueljoli** [comment]
@osukaa Are you opposed to removing `package-lock.json`?

**Comment by Samueljoli** [comment]
Linting issues are related to `hapi-no-var` rule not being found. Looking at the travis logs, I'm seeing that `esling-plugin-hapi` isn't pulling in that dep. But I don't see any reason why it shouldn't be... 🤔

**Comment by Marsup** [comment]
The package-lock should never have made it into the repo.

**Comment by osukaa** [comment]
I agree with @Marsup, the package-lock shouldn't be in the repo

File: .travis.yml [hunk 1]
```diff
@@ -1,8 +1,6 @@
 language: node_js
 
 node_js:
-  - "lts/argon"
-  - "lts/boron"
-  - "node"
+  - "8"
```
Comment (mrlannigan) [inline .travis.yml]: You may want to follow hapi's lead in the versions that should be tested against. See here: https://github.com/hapijs/hapi/blob/d3bbf3ac5955bcfc7ff6dc770afff79ca8c95bd8/.travis.yml#L3-L6

**Comment by Samueljoli** [comment]
@mrlannigan Updated. @Marsup @osukaa `package-lock` toasted.

**Comment by osukaa** [comment]
awesome

**Comment by osukaa** [comment]
Please update the README :)

File: lib/index.js [hunk 2]
```diff
@@ -50,43 +50,40 @@ internals.schema = Joi.object({
...
 
-    return function (request, reply) {
-
-        settings.mapUri(request, (err, uri, headers) => {
+    return async function (request, h) {
 
-            if (err) {
-                return reply(err);
-            }
+        try {
```
Comment (Marsup) [inline lib/index.js]: The try/catch is completely useless if the only thing it does is rethrow.

File: lib/index.js [hunk 3]
```diff
@@ -104,17 +101,17 @@ internals.handler = function (route, handlerOptions) {
                 options.headers = Hoek.clone(request.headers);
                 delete options.headers.host;
 
-                if (settings.acceptEncoding === false) {                    // Defaults to true
+                if (settings.acceptEncoding === false) { // Defaults to true
                     delete options.headers['accept-encoding'];
                 }
 
-                if (options.headers.cookie) {
+                if (options.headers.cookie) { //TODO: Add documentation for server.states.passThrough
```
Comment (Marsup) [inline lib/index.js]: Is there an issue about this rather than this TODO ?

**Comment by Samueljoli** [comment]
@spanditcaa @osukaa I was holding off on updating the ReadMe until the PR has been fully reviewed, mainly around the breaking change to `mapUri`. But if we can all agree on that change then I'll update the readme.

**Comment by spanditcaa** [comment]
@Samueljoli what if I pass in my own mapUri that does some async lookup?  Perhaps ````const { uri, headers } = await settings.mapUri(request); ```` ?

**Comment by Samueljoli** [comment]
@spanditcaa Yeah that's a good idea, 

```js
let uri, headers;

try {
  await settings.mapUri(request);
}
catch(e) {
  //...would like to reply error to client. I believe throwing would achieve this?
  throw e;
}
```

File: lib/index.js [hunk 4]
```diff
@@ -50,138 +50,133 @@ internals.schema = Joi.object({
...
 
     if (settings.ttl === 'upstream') {
         settings._upstreamTtl = true;
     }
 
-    return function (request, reply) {
+    return async function (request, h) {
 
-        settings.mapUri(request, (err, uri, headers) => {
+        const { uri, headers } = settings.mapUri(request);
```
Comment (spanditcaa) [inline lib/index.js]: await settings.mapUri(request) here, to support async custom mapUri functions

**Comment by Samueljoli** [comment]
@Marsup 😁 haha, yeah a reminder for me to open an issue.

**Comment by Samueljoli** [comment]
@spanditcaa @osukaa Readme has been updated 😄 .

**Comment by pthrasher** [comment]
@Samueljoli @Marsup @spanditcaa @osukaa Any chance we can push to get this merged soon? This is the last piece I need for v17 upgrade in my project.

**Comment by spanditcaa** [comment]
@osukaa @Marsup any further comments?

File: package.json [hunk 5]
```diff
@@ -12,19 +12,19 @@
     "plugin"
   ],
   "engines": {
-    "node": ">=4.0.0"
+    "node": ">=8.x.x"
   },
   "dependencies": {
     "boom": "5.x.x",
```
Comment (Marsup) [inline package.json]: Please bump all dependencies to latest, this doesn't make sense to stick to pre-v17 ones.

File: test/index.js [hunk 6]
```diff
@@ -34,81 +34,89 @@ describe('H2o2', () => {
...
+    const provisionServer = async function (options) {
+
+        const server = Hapi.server(options);
+
+        try {
+            await server.register(H2o2);
+        }
+        catch (err) {
+            console.log(err);
+            process.exit(1);
```
Comment (Marsup) [inline test/index.js]: Exit ? Why ?

**Comment by Samueljoli** [comment]
@Marsup some clarification please 😄.

> The proxy handler doesn't exist anymore, since hapi removed custom handlers.

I'm assuming you mean the decoration change of: `Move server.handler() to use server.decorate() instead.`?

And by merging with the `h` method you mean [this](https://github.com/Samueljoli/h2o2/tree/hapiv17#usage) which as you can see I've already updated the documentation.

File: package.json (hunk 5 above)
Comment (Samueljoli) [reply package.json]: @Marsup Deps updated

**Comment by Marsup** [comment]
There are still many examples like :
```js
server.route({
  method: 'GET',
  handler: {
    proxy: {
      ...
    }
  }
});
```
This is not allowed anymore. All the times `proxy handler` is mentioned it seems like an error as it refers to this old world.

**Comment by Samueljoli** [comment]
@marsup 
```js
server.route({
  method: 'GET',
  handler: {
    proxy: {
      ...
    }
  }
});
```
Per the documentation, it seems like [this](https://github.com/hapijs/hapi/blob/master/API.md#-serverdecoratetype-property-method-options) is still supported. Right? I would imagine that I'd run into this in the form of a broken test. But that wasn't the case.

**Comment by Marsup** [comment]
OK my bad, I was convinced Eran deprecated it.

**Comment by spanditcaa** [comment]
@Samueljoli @pthrasher -- I'm going to merge the other PRs that are useful to hapi v<17 users ahead of this so we have a final v6 release; then will take care of any merge issues to get 7.0 out.  I've spent more time in front of the stove than the computer in the last few days, but should get this out by Monday.

**Comment by spanditcaa** [comment]
@Samueljoli I resolved the merge conflicts and merged this via command line -- i'm not sure why the PR didn't close but I pushed up the updates as h2o2 v7.0.0

**Comment by Samueljoli** [comment]
@spanditcaa all good! But I did lose my contribution 😞

**Comment by WesTyler** [comment]
> i'm not sure why the PR didn't close

looks like you did a rebase to squash the commits or something? None of the commit history from this branch/PR show up. If that's the case, then that's why it didn't merge/close the PR.

**Comment by WesTyler** [comment]
Also as a point of interest I do not see either a `7.0.0` commit in the history or a `v7.0.0` tag in the Releases.

**Comment by WesTyler** [comment]
I'm actually really interested in how you did the merge. I'm not sure how you could merge the changes, bump the npm version, and deploy the new package with such an empty commit history :P

**Comment by WesTyler** [comment]
@spanditcaa okay no really, how did you do the merge? 

I think it's really important to ensure that new contributors to the project feel appreciated, and even if we cannot recover the git history to credit @Samueljoli I want to make sure this doesn't happen the next time someone opens a PR here.

**Comment by spanditcaa** [comment]
Here were the steps @WesTyler - substitute @Samueljoli branch/repo.  My apologies again for the trouble @Samueljoli.

````
git checkout -b spanditcaa-master master
git pull https://github.com/spanditcaa/h2o2.git master

<addressed merge conflict>
git commit -a -m "addressed merge conflict"
npm version major

git checkout master
git merge --no-ff spanditcaa-master
git push origin master
 ````

**Pull request closed without merging**